pixeltable/env.py

```diff
@@ -416,25 +393,62 @@ def _set_up(self, echo: bool = False, reinit_db: bool = False) -> None:
-        if create_db or db_connect_str is not None:
-            from pixeltable import metadata
-
-            metadata.schema.base_metadata.create_all(self._sa_engine)
-            metadata.create_system_info(self._sa_engine)
+        # Create catalog tables and system metadata
+        self._init_metadata()
 
         self.console_logger.info(f'Connected to Pixeltable database at: {self.db_url}')
 
         # we now have a home directory and db; start other services
         self._set_up_runtime()
         self.log_to_stdout(False)
 
-    @property
-    def default_system_db_url(self):
-        if self._db_server is None:
-            return self._dbms.default_system_db_url()
+    def _init_db(self, config: Config) -> None:
+        """
+        Initialize the pixeltable database along with its associated DBMS.
+        """
+        db_connect_str = config.get_string_value('DB_CONNECT_STR')
+        if db_connect_str is not None:
+            try:
+                db_url = sql.make_url(db_connect_str)
+            except sql.exc.ArgumentError as e:
+                error = f'Invalid db connection string {db_connect_str}: {e}'
+                self._logger.error(error)
+                raise excs.Error(error) from e
+            self._db_url = db_url.render_as_string(hide_password=False)
+            self._db_name = db_url.database  # use the dbname given in connect string
+            dialect = db_url.get_dialect().name
+            if dialect == 'cockroachdb':
+                self._dbms = CockroachDbms(db_url)
+            else:
+                raise excs.Error(f'Unsupported DBMS {dialect}')
+            # Check if database exists
+            if not self._store_db_exists():
+                error = f'Database {self._db_name!r} does not exist'
+                self._logger.error(error)
+                raise excs.Error(error)
+            self._logger.info(f'Using database at: {self.db_url}')
         else:
-            return self._db_server.get_uri(database='postgres', driver='psycopg')
+            self._db_name = os.environ.get('PIXELTABLE_DB', 'pixeltable')
+            self._pgdata_dir = Path(os.environ.get('PIXELTABLE_PGDATA', str(Config.get().home / 'pgdata')))
+            # cleanup_mode=None will leave the postgres process running after Python exits
+            # cleanup_mode='stop' will terminate the postgres process when Python exits
+            # On Windows, we need cleanup_mode='stop' because child processes are killed automatically when the parent
+            # process (such as Terminal or VSCode) exits, potentially leaving it in an unusable state.
+            cleanup_mode = 'stop' if platform.system() == 'Windows' else None
+            self._db_server = pixeltable_pgserver.get_server(self._pgdata_dir, cleanup_mode=cleanup_mode)
+            self._db_url = self._db_server.get_uri(database=self._db_name, driver='psycopg')
+            self._dbms = PostgresqlDbms(sql.make_url(self._db_url))
```

Comment:
Before you return, add a few asserts like self._dbms is not None. That'll also help the reader see what needs to get set here.